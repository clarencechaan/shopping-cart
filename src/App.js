import { BrowserRouter, Routes, Route } from "react-router-dom";
import NavBar from "./components/NavBar";
import Home from "./components/Home";
import MusicShop from "./components/MusicShop";
import MerchShop from "./components/MerchShop";
import SaleShop from "./components/SaleShop";
import Cart from "./components/Cart";
import Details from "./components/Details";
import items from "./items/items.js";
import "./styles/App.css";
import { useState, useEffect } from "react";

const App = () => {
  const [shopItems, setShopItems] = useState([]);
  const [cart, setCart] = useState([]);

  useEffect(() => setShopItems(items), []);

  function handleAddToCart(item, qty) {
    if (!qty) return;
    if (!isItemInCart(item, cart)) {
      setCart((prevCart) => {
        const newCartItem = { ...item, quantity: qty };
        return [...prevCart, newCartItem];
      });
    } else {
      setCart((prevCart) => {
        const prevCartItem = prevCart.find((c) => c.id === item.id);
        const newCartItem = {
          ...prevCartItem,
          quantity: prevCartItem.quantity + qty,
        };
        const index = prevCart.findIndex((c) => c.id === item.id);
        return [
          ...prevCart.slice(0, index),
          newCartItem,
          ...prevCart.slice(index + 1),
        ];
      });
    }
    removeZeroQuantity();
  }

  function removeZeroQuantity() {
    setCart((prevCart) => prevCart.filter((item) => item.quantity > 0));
  }

  function isItemInCart(item) {
    for (const c of cart) {
      if (c.id === item.id) return true;
    }
    return false;
  }

  function getQtyInCart(id) {
    if (isItemInCart({ id: id })) {
      return cart.find((c) => c.id === id).quantity;
    } else {
      return 0;
    }
  }

  function getTotalQtyInCart() {
    let total = 0;
    for (const c of cart) {
      total += c.quantity;
    }
    return total;
  }

  // function addQty(id, qty) {
  //   setCart((prevCart) => {
  //     const prevCartItem = prevCart.find((c) => c.id === id);
  //     const newCartItem = {
  //       ...prevCartItem,
  //       quantity: prevCartItem.quantity + 1,
  //     };
  //     const index = prevCart.findIndex((c) => c.id === id);
  //     return [
  //       ...prevCart.slice(0, index),
  //       newCartItem,
  //       ...prevCart.slice(index + 1),
  //     ];
  //   });
  // }

  console.log(cart);

  return (
    <BrowserRouter>
      <NavBar totalQtyInCart={getTotalQtyInCart()} />
      <Routes>
        <Route path="/" element={<Home />} />
        <Route
          path="/music-shop"
          element={<MusicShop shopItems={shopItems} />}
        />
        <Route
          path="/merch-shop"
          element={<MerchShop shopItems={shopItems} />}
        />
        <Route path="/sale-shop" element={<SaleShop shopItems={shopItems} />} />
        <Route
          path="/cart"
          element={
            <Cart
              cartItems={cart}
              getTotalQtyInCart={getTotalQtyInCart}
              handleAddToCart={handleAddToCart}
            />
          }
        />
        <Route
          path="/details"
          element={
            <Details
              handleAddToCart={handleAddToCart}
              getQtyInCart={getQtyInCart}
            />
          }
        />
      </Routes>
    </BrowserRouter>
  );
};

export default App;
