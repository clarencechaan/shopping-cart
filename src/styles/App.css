body {
  margin: 0;
}

#root {
  width: 100vw;
  height: 100vh;
  max-width: 100%;
  font-family: "Pavanam", sans-serif;
  display: flex;
  flex-direction: column;
  align-items: center;
}

#root > div:last-child {
  flex: 1;
}

.home {
  width: 100%;
  height: 100%;
}

.music-shop,
.merch-shop,
.sale-shop,
.cart {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  width: 1192px;
  padding: 0 24px 24px 24px;
}

.cart {
  width: 1096px;
}
