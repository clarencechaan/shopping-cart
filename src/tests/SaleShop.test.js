import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";

describe("SaleShop component", () => {
  it("renders SaleShop", () => {
    const { container } = render(<App />);
    const sale = screen.getByText(/^SALE$/);
    userEvent.click(sale);

    expect(container).toMatchSnapshot();
  });

  it("ShopItem link", () => {
    const { container } = render(<App />);
    const img = screen.getByAltText("Take Care by Drake");
    userEvent.click(img);

    expect(container).toMatchSnapshot();
  });
});
