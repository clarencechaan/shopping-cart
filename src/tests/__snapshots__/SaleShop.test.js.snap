// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaleShop component ShopItem link 1`] = `
<div>
  <nav
    class="nav-bar"
  >
    <a
      class="logo"
      href="/shopping-cart"
    >
      auditty
    </a>
    <div>
      <a
        class="nav-link"
        href="/shopping-cart"
      >
        HOME
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/music-shop"
      >
        MUSIC
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/merch-shop"
      >
        MERCH
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/sale-shop"
        style="color: red;"
      >
        SALE
      </a>
    </div>
    <a
      class="cart-logo"
      href="/shopping-cart/cart"
    >
      <img
        alt="cart"
        src="shopping-cart.png"
      />
      <div
        class="cart-badge hidden"
      />
    </a>
  </nav>
  <div
    class="details"
  >
    <img
      alt=""
      class="details-img"
      src="take-care.jpg"
    />
    <div
      class="details-info"
    >
      <div
        class="qty-in-cart"
      />
      <div
        class="details-title"
      >
        Take Care by Drake
      </div>
      <div
        class="details-price"
      >
        $
        9.99
      </div>
      <div
        class="quantity css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-1s2u09g-control"
        >
          <div
            class=" css-319lph-ValueContainer"
          >
            <div
              class=" css-qc6sy-singleValue"
            >
              1
            </div>
            <div
              class=" css-6j8wv5-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-controls="react-select-2-listbox"
                aria-expanded="false"
                aria-haspopup="true"
                aria-owns="react-select-2-listbox"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-tlfecz-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <button
        class="add-to-cart active"
      >
        ADD TO CART
      </button>
    </div>
  </div>
</div>
`;

exports[`SaleShop component renders SaleShop 1`] = `
<div>
  <nav
    class="nav-bar"
  >
    <a
      class="logo"
      href="/shopping-cart"
    >
      auditty
    </a>
    <div>
      <a
        class="nav-link"
        href="/shopping-cart"
      >
        HOME
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/music-shop"
      >
        MUSIC
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/merch-shop"
      >
        MERCH
      </a>
      <a
        class="nav-link"
        href="/shopping-cart/sale-shop"
        style="color: red;"
      >
        SALE
      </a>
    </div>
    <a
      class="cart-logo"
      href="/shopping-cart/cart"
    >
      <img
        alt="cart"
        src="shopping-cart.png"
      />
      <div
        class="cart-badge hidden"
      />
    </a>
  </nav>
  <div
    class="sale-shop"
  >
    <h2>
      Sale
    </h2>
    <div
      class="items-container"
    >
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Take Care by Drake"
            class="shop-item-img"
            src="take-care.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Take Care by Drake
        </div>
        <div
          class="shop-price"
        >
          $
          9.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Care for Me by Saba"
            class="shop-item-img"
            src="care.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Care for Me by Saba
        </div>
        <div
          class="shop-price"
        >
          $
          9.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Food & Liquor by Lupe Fiasco"
            class="shop-item-img"
            src="food-liquor.jpg"
          />
        </a>
        <div
          class="shop-title"
        >
          Food & Liquor by Lupe Fiasco
        </div>
        <div
          class="shop-price"
        >
          $
          9.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="OVO Essentials Hoodie"
            class="shop-item-img"
            src="ovo-hoodie.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          OVO Essentials Hoodie
        </div>
        <div
          class="shop-price"
        >
          $
          79.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="ISSA Money Pool Tee"
            class="shop-item-img"
            src="issa-tee.jpeg"
          />
        </a>
        <div
          class="shop-title"
        >
          ISSA Money Pool Tee
        </div>
        <div
          class="shop-price"
        >
          $
          24.99
        </div>
      </div>
      <div
        class="shop-item"
      >
        <a
          href="/shopping-cart/details"
        >
          <img
            alt="Enjoy Life Tee"
            class="shop-item-img"
            src="6lack-tee.webp"
          />
        </a>
        <div
          class="shop-title"
        >
          Enjoy Life Tee
        </div>
        <div
          class="shop-price"
        >
          $
          24.99
        </div>
      </div>
    </div>
  </div>
</div>
`;
